`upload_benchmark_results` workflow hangs on `get_dataset_details` due to full dataset downloads

### Environment Details

* SDGym version: 0.14.3

### Error Description

When running the benchmark upload script for multi_table modality, the process silently hangs after for building the `Dataset_Details.xlsx` and never completes. On Github [action](https://github.com/sdv-dev/SDGym/actions/runs/25485541481/job/74780048263) this causes a:
`Error: Process completed with exit code 143.`

The root cause lies in `get_dataset_details`: it calls `explorer.summarize_datasets(modality=modality)` on both the `public` and `private` buckets, which ultimately downloads the data ZIP for every dataset in each bucket (not just those in the current benchmark run) to compute the `Total_Num_Rows` column. Now that the `rel-bench` datasets have been added to the private bucket, the process fails when downloading them.

### Expected behavior

`get_dataset_details` should only fetch information for datasets that were part of the benchmark run. To do this:

- Add a datasets parameter to `DatasetExplorer._load_and_summarize_datasets()`: a list of dataset names. When None (the default), all datasets in the bucket are summarized. When provided, only the datasets in the list are summarized.
- Update `get_dataset_details` to call `_load_and_summarize_datasets` passing only the datasets present in the benchmark results.

### Additional Context

`Total_Num_Rows` is currently the only field in the dataset details table that requires downloading the data ZIP. This is out of scope for this issue, but we might consider storing this information elsewhere (e.g., in the dataset's metainfo.yaml) so we don't need to download the data to get it.

--- a/sdgym/dataset_explorer.py
+++ b/sdgym/dataset_explorer.py
@@ -212,12 +212,16 @@ class DatasetExplorer:
 
         return data_summary
 
-    def _load_and_summarize_datasets(self, modality):
+    def _load_and_summarize_datasets(self, modality, datasets=None):
         """Load all datasets for the given modality and compute summary statistics.
 
         Args:
             modality (str):
                 The dataset modality to load (e.g., 'single_table' or 'multi_table').
+            datasets (list[str], optional):
+                A list of dataset names to restrict the summarization to. If ``None``,
+                all available datasets for the given modality are summarized.
+                Defaults to ``None``.
 
         Returns:
             list[dict]:
@@ -226,12 +230,17 @@ class DatasetExplorer:
         """
         results = []
 
-        datasets = _get_available_datasets(
+        available_datasets = _get_available_datasets(
             modality=modality,
             bucket=self._bucket_name,
             s3_client=self.s3_client,
         )
-        for _, dataset_row in datasets.iterrows():
+        if datasets is not None:
+            available_datasets = available_datasets[
+                available_datasets['dataset_name'].isin(datasets)
+            ]
+
+        for _, dataset_row in available_datasets.iterrows():
             dataset_name = dataset_row['dataset_name']
             dataset_size_mb = dataset_row['size_MB']
             dataset_num_table = dataset_row['num_tables']

--- a/sdgym/result_explorer/result_handler.py
+++ b/sdgym/result_explorer/result_handler.py
@@ -178,8 +178,18 @@ class ResultsHandler(ABC):
             .str.replace(REGEX_SYNTHESIZER_NAME, '', regex=True)
             .str.strip()
         )
+        if 'Error' in aggregated_results.columns:
+            is_uniform = aggregated_results['Synthesizer'].str.contains('UniformSynthesizer')
+            has_error = aggregated_results['Error'].notna()
+            aggregated_results['Successful_Uniform'] = ~(is_uniform & has_error)
+            aggregated_results = aggregated_results.sort_values(
+                'Successful_Uniform', ascending=False, kind='stable'
+            )
+            aggregated_results = aggregated_results.drop(columns='Successful_Uniform')
+
         aggregated_results = aggregated_results.drop_duplicates(
-            subset=['Dataset', 'Synthesizer'], keep='first'
+            subset=['Dataset', 'Synthesizer'],
+            keep='first',
         )
         aggregated_results = _add_adjusted_scores(aggregated_results, timeout=TIMEOUT)
 

--- a/sdgym/result_explorer/result_handler.py
+++ b/sdgym/result_explorer/result_handler.py
@@ -188,7 +188,8 @@ class ResultsHandler(ABC):
             aggregated_results = aggregated_results.drop(columns='Successful_Uniform')
 
         aggregated_results = aggregated_results.drop_duplicates(
-            subset=['Dataset', 'Synthesizer'], keep='first',
+            subset=['Dataset', 'Synthesizer'],
+            keep='first',
         )
         aggregated_results = _add_adjusted_scores(aggregated_results, timeout=TIMEOUT)
 

--- a/sdgym/result_explorer/result_handler.py
+++ b/sdgym/result_explorer/result_handler.py
@@ -188,8 +188,7 @@ class ResultsHandler(ABC):
             aggregated_results = aggregated_results.drop(columns='Successful_Uniform')
 
         aggregated_results = aggregated_results.drop_duplicates(
-            subset=['Dataset', 'Synthesizer'],
-            keep='first',
+            subset=['Dataset', 'Synthesizer'], keep='first',
         )
         aggregated_results = _add_adjusted_scores(aggregated_results, timeout=TIMEOUT)
 

--- a/sdgym/run_benchmark/upload_benchmark_results.py
+++ b/sdgym/run_benchmark/upload_benchmark_results.py
@@ -167,7 +167,7 @@ def get_dataset_details(results, modality, aws_access_key_id, aws_secret_access_
         `pd.DataFrame`:
             Dataset details DataFrame.
     """
-    datasets = results['Dataset'].unique().tolist()
+    dataset_list = results['Dataset'].unique().tolist()
     explorers = {
         'Public': DatasetExplorer(
             s3_url=SDV_DATASETS_PUBLIC_BUCKET,
@@ -182,7 +182,7 @@ def get_dataset_details(results, modality, aws_access_key_id, aws_secret_access_
     }
 
     dataset_infos = []
-    remaining_datasets = set(datasets)
+    remaining_datasets = set(dataset_list)
     for availability, explorer in explorers.items():
         summary = explorer._load_and_summarize_datasets(
             modality=modality, datasets=list(remaining_datasets)

--- a/sdgym/run_benchmark/upload_benchmark_results.py
+++ b/sdgym/run_benchmark/upload_benchmark_results.py
@@ -18,12 +18,7 @@ from pydrive2.auth import GoogleAuth
 from pydrive2.drive import GoogleDrive
 
 from sdgym import DatasetExplorer
-from sdgym.datasets import (
-    SDV_DATASETS_PRIVATE_BUCKET,
-    SDV_DATASETS_PUBLIC_BUCKET,
-    _get_available_datasets,
-    _load_dataset_with_client,
-)
+from sdgym.datasets import SDV_DATASETS_PRIVATE_BUCKET, SDV_DATASETS_PUBLIC_BUCKET
 from sdgym.result_explorer.result_explorer import ResultsExplorer
 from sdgym.result_writer import LocalResultsWriter
 from sdgym.run_benchmark.utils import (
@@ -172,8 +167,7 @@ def get_dataset_details(results, modality, aws_access_key_id, aws_secret_access_
         `pd.DataFrame`:
             Dataset details DataFrame.
     """
-    dataset_list = results['Dataset'].unique().tolist()
-    datasets_to_match = set(dataset_list)
+    datasets = results['Dataset'].unique().tolist()
     explorers = {
         'Public': DatasetExplorer(
             s3_url=SDV_DATASETS_PUBLIC_BUCKET,
@@ -188,45 +182,23 @@ def get_dataset_details(results, modality, aws_access_key_id, aws_secret_access_
     }
 
     dataset_infos = []
+    remaining_datasets = set(datasets)
     for availability, explorer in explorers.items():
-        available_datasets = _get_available_datasets(
-            modality=modality,
-            bucket=explorer._bucket_name,
-            s3_client=explorer.s3_client,
+        summary = explorer._load_and_summarize_datasets(
+            modality=modality, datasets=list(remaining_datasets)
         )
-        available_datasets = available_datasets[
-            available_datasets['dataset_name'].isin(datasets_to_match)
-        ]
-        if available_datasets.empty:
+        if not summary:
             continue
 
-        for _, row in available_datasets.iterrows():
-            dataset_name = row['dataset_name']
-            data, metadata_dict = _load_dataset_with_client(
-                modality=modality,
-                dataset=dataset_name,
-                bucket=explorer._bucket_name,
-                s3_client=explorer.s3_client,
-            )
-            metadata_summary = DatasetExplorer.get_metadata_summary(metadata_dict)
-            data_summary = DatasetExplorer.get_data_summary(data)
-            dataset_infos.append({
-                'Dataset': dataset_name,
-                'Datasize_Size_MB': row['size_MB'],
-                'Num_Tables': row['num_tables'],
-                'Total_Num_Columns': metadata_summary['Total_Num_Columns'],
-                'Total_Num_Columns_Categorical': metadata_summary['Total_Num_Columns_Categorical'],
-                'Total_Num_Columns_Numerical': metadata_summary['Total_Num_Columns_Numerical'],
-                'Total_Num_Rows': data_summary['Total_Num_Rows'],
-                'Num_Relationships': metadata_summary['Num_Relationships'],
-                'Max_Schema_Depth': metadata_summary['Max_Schema_Depth'],
-                'Availability': availability,
-            })
+        summary_df = pd.DataFrame(summary)
+        summary_df['Availability'] = availability
+        dataset_infos.append(summary_df)
+        remaining_datasets -= set(summary_df['Dataset'])
 
     if not dataset_infos:
         return pd.DataFrame(columns=DATASET_DETAILS_COLUMNS)
 
-    dataset_infos = pd.DataFrame(dataset_infos)
+    dataset_infos = pd.concat(dataset_infos, ignore_index=True)
     best_model_map = (
         results
         .sort_values('Adjusted_Quality_Score', ascending=False)

--- a/sdgym/run_benchmark/upload_benchmark_results.py
+++ b/sdgym/run_benchmark/upload_benchmark_results.py
@@ -18,7 +18,12 @@ from pydrive2.auth import GoogleAuth
 from pydrive2.drive import GoogleDrive
 
 from sdgym import DatasetExplorer
-from sdgym.datasets import SDV_DATASETS_PRIVATE_BUCKET, SDV_DATASETS_PUBLIC_BUCKET
+from sdgym.datasets import (
+    SDV_DATASETS_PRIVATE_BUCKET,
+    SDV_DATASETS_PUBLIC_BUCKET,
+    _get_available_datasets,
+    _load_dataset_with_client,
+)
 from sdgym.result_explorer.result_explorer import ResultsExplorer
 from sdgym.result_writer import LocalResultsWriter
 from sdgym.run_benchmark.utils import (
@@ -168,6 +173,7 @@ def get_dataset_details(results, modality, aws_access_key_id, aws_secret_access_
             Dataset details DataFrame.
     """
     dataset_list = results['Dataset'].unique().tolist()
+    datasets_to_match = set(dataset_list)
     explorers = {
         'Public': DatasetExplorer(
             s3_url=SDV_DATASETS_PUBLIC_BUCKET,
@@ -182,23 +188,45 @@ def get_dataset_details(results, modality, aws_access_key_id, aws_secret_access_
     }
 
     dataset_infos = []
-    remaining_datasets = set(dataset_list)
     for availability, explorer in explorers.items():
-        summary = explorer.summarize_datasets(modality=modality)
-        summary = (
-            summary.set_index('Dataset').reindex(remaining_datasets).dropna(how='all').reset_index()
+        available_datasets = _get_available_datasets(
+            modality=modality,
+            bucket=explorer._bucket_name,
+            s3_client=explorer.s3_client,
         )
-        if summary.empty:
+        available_datasets = available_datasets[
+            available_datasets['dataset_name'].isin(datasets_to_match)
+        ]
+        if available_datasets.empty:
             continue
 
-        summary['Availability'] = availability
-        dataset_infos.append(summary)
-        remaining_datasets -= set(summary['Dataset'])
+        for _, row in available_datasets.iterrows():
+            dataset_name = row['dataset_name']
+            data, metadata_dict = _load_dataset_with_client(
+                modality=modality,
+                dataset=dataset_name,
+                bucket=explorer._bucket_name,
+                s3_client=explorer.s3_client,
+            )
+            metadata_summary = DatasetExplorer.get_metadata_summary(metadata_dict)
+            data_summary = DatasetExplorer.get_data_summary(data)
+            dataset_infos.append({
+                'Dataset': dataset_name,
+                'Datasize_Size_MB': row['size_MB'],
+                'Num_Tables': row['num_tables'],
+                'Total_Num_Columns': metadata_summary['Total_Num_Columns'],
+                'Total_Num_Columns_Categorical': metadata_summary['Total_Num_Columns_Categorical'],
+                'Total_Num_Columns_Numerical': metadata_summary['Total_Num_Columns_Numerical'],
+                'Total_Num_Rows': data_summary['Total_Num_Rows'],
+                'Num_Relationships': metadata_summary['Num_Relationships'],
+                'Max_Schema_Depth': metadata_summary['Max_Schema_Depth'],
+                'Availability': availability,
+            })
 
     if not dataset_infos:
         return pd.DataFrame(columns=DATASET_DETAILS_COLUMNS)
 
-    dataset_infos = pd.concat(dataset_infos, ignore_index=True)
+    dataset_infos = pd.DataFrame(dataset_infos)
     best_model_map = (
         results
         .sort_values('Adjusted_Quality_Score', ascending=False)

--- a/tests/unit/result_explorer/test_result_handler.py
+++ b/tests/unit/result_explorer/test_result_handler.py
@@ -361,6 +361,36 @@ class TestResultsHandler:
         pd.testing.assert_frame_equal(processed_results, expected_results, check_dtype=False)
         assert missing_columns == []
 
+    def test__process_results_keeps_successful_uniform_row(self):
+        """Test that the `_process_results` method keeps the successful UniformSynthesizer row."""
+        results = [
+            pd.DataFrame({
+                'Dataset': ['A', 'A', 'A', 'A'],
+                'Synthesizer': [
+                    'MultiTableUniformSynthesizer',
+                    'MultiTableUniformSynthesizer(1)',
+                    'Synth1',
+                    'Synth2',
+                ],
+                'Train_Time': [10, 20, 5, 6],
+                'Sample_Time': [1, 2, 1, 1],
+                'Quality_Score': [0.1, 0.2, 0.3, 0.4],
+                'Error': ['Synthesizer Timeout', None, None, None],
+                'Adjusted_Total_Time': [11, 22, 6, 7],
+                'Adjusted_Quality_Score': [0.1, 0.2, 0.3, 0.4],
+            })
+        ]
+        handler = Mock()
+
+        processed_results, _ = ResultsHandler._process_results(handler, results)
+
+        uniform_row = processed_results.loc[
+            processed_results['Synthesizer'] == 'MultiTableUniformSynthesizer'
+        ].iloc[0]
+        assert pd.isna(uniform_row['Error'])
+        assert uniform_row['Train_Time'] == 20
+        assert uniform_row['Sample_Time'] == 2
+
     def test_summarize(self):
         """Test the `summarize` method."""
         # Setup

--- a/tests/unit/run_benchmark/test_upload_benchmark_result.py
+++ b/tests/unit/run_benchmark/test_upload_benchmark_result.py
@@ -170,11 +170,13 @@ def test_upload_to_drive_file_not_found(tmp_path):
 
 
 @patch('sdgym.run_benchmark.upload_benchmark_results.DatasetExplorer')
+@patch('sdgym.run_benchmark.upload_benchmark_results._load_dataset_with_client')
+@patch('sdgym.run_benchmark.upload_benchmark_results._get_available_datasets')
 @patch(
     'sdgym.run_benchmark.upload_benchmark_results.DATASET_DETAILS_COLUMNS',
-    ['Dataset', 'Rows', 'Availability', 'Best Model', 'Data Modality'],
+    ['Dataset', 'Total_Num_Rows', 'Availability', 'Best Model', 'Data Modality'],
 )
-def test_get_dataset_details(mock_dataset_explorer):
+def test_get_dataset_details(mock_get_available_datasets, mock_load_dataset, mock_dataset_explorer):
     """Test the `get_dataset_details` method"""
     # Setup
     aws_access_key_id = 'access'
@@ -197,21 +199,76 @@ def test_get_dataset_details(mock_dataset_explorer):
     private_explorer = Mock()
     mock_dataset_explorer.side_effect = [public_explorer, private_explorer]
 
-    public_explorer.summarize_datasets.return_value = pd.DataFrame({
-        'Dataset': ['A', 'B'],
-        'Rows': [10, 20],
-    })
-    private_explorer.summarize_datasets.return_value = pd.DataFrame({
-        'Dataset': ['C'],
-        'Rows': [30],
-    })
+    public_explorer._bucket_name = SDV_DATASETS_PUBLIC_BUCKET
+    public_explorer.s3_client = Mock()
+    private_explorer._bucket_name = SDV_DATASETS_PRIVATE_BUCKET
+    private_explorer.s3_client = Mock()
 
-    # Run
-    dataset_details = get_dataset_details(
-        results, modality, aws_access_key_id, aws_secret_access_key
-    )
+    mock_get_available_datasets.side_effect = [
+        pd.DataFrame({
+            'dataset_name': ['A', 'B', 'X'],
+            'size_MB': [1.0, 2.0, 999.0],
+            'num_tables': [1, 1, 1],
+        }),
+        pd.DataFrame({
+            'dataset_name': ['C'],
+            'size_MB': [3.0],
+            'num_tables': [1],
+        }),
+    ]
+    mock_load_dataset.side_effect = [
+        ({'table': pd.DataFrame({'col': [1] * 10})}, {'metadata': 1}),
+        ({'table': pd.DataFrame({'col': [1] * 20})}, {'metadata': 1}),
+        ({'table': pd.DataFrame({'col': [1] * 30})}, {'metadata': 1}),
+    ]
+
+    with (
+        patch(
+            'sdgym.run_benchmark.upload_benchmark_results.DatasetExplorer.get_metadata_summary',
+            side_effect=[
+                {
+                    'Total_Num_Columns': 1,
+                    'Total_Num_Columns_Categorical': 1,
+                    'Total_Num_Columns_Numerical': 0,
+                    'Num_Relationships': 0,
+                    'Max_Schema_Depth': 1,
+                },
+                {
+                    'Total_Num_Columns': 1,
+                    'Total_Num_Columns_Categorical': 1,
+                    'Total_Num_Columns_Numerical': 0,
+                    'Num_Relationships': 0,
+                    'Max_Schema_Depth': 1,
+                },
+                {
+                    'Total_Num_Columns': 1,
+                    'Total_Num_Columns_Categorical': 1,
+                    'Total_Num_Columns_Numerical': 0,
+                    'Num_Relationships': 0,
+                    'Max_Schema_Depth': 1,
+                },
+            ],
+        ),
+        patch(
+            'sdgym.run_benchmark.upload_benchmark_results.DatasetExplorer.get_data_summary',
+            side_effect=[{'Total_Num_Rows': 10}, {'Total_Num_Rows': 20}, {'Total_Num_Rows': 30}],
+        ),
+    ):
+        # Run
+        dataset_details = get_dataset_details(
+            results, modality, aws_access_key_id, aws_secret_access_key
+        )
 
     # Assert
+    assert mock_get_available_datasets.call_count == 2
+    assert mock_load_dataset.call_count == 3
+    expected = pd.DataFrame({
+        'Dataset': ['A', 'B', 'C'],
+        'Total_Num_Rows': [10, 20, 30],
+        'Availability': ['Public', 'Public', 'Private'],
+        'Best Model': ['TVAESynthesizer', 'GaussianCopulaSynthesizer', 'CTGANSynthesizer'],
+        'Data Modality': [modality, modality, modality],
+    })
     assert mock_dataset_explorer.call_count == 2
     for _call, bucket in zip(mock_dataset_explorer.call_args_list, dataset_bucket):
         assert _call.kwargs == {
@@ -219,16 +276,6 @@ def test_get_dataset_details(mock_dataset_explorer):
             'aws_access_key_id': aws_access_key_id,
             'aws_secret_access_key': aws_secret_access_key,
         }
-
-    public_explorer.summarize_datasets.assert_called_once_with(modality=modality)
-    private_explorer.summarize_datasets.assert_called_once_with(modality=modality)
-    expected = pd.DataFrame({
-        'Dataset': ['A', 'B', 'C'],
-        'Rows': [10, 20, 30],
-        'Availability': ['Public', 'Public', 'Private'],
-        'Best Model': ['TVAESynthesizer', 'GaussianCopulaSynthesizer', 'CTGANSynthesizer'],
-        'Data Modality': [modality, modality, modality],
-    })
     assert_frame_equal(
         dataset_details.sort_values('Dataset').reset_index(drop=True),
         expected.sort_values('Dataset').reset_index(drop=True),
@@ -237,11 +284,14 @@ def test_get_dataset_details(mock_dataset_explorer):
 
 
 @patch('sdgym.run_benchmark.upload_benchmark_results.DatasetExplorer')
+@patch('sdgym.run_benchmark.upload_benchmark_results._get_available_datasets')
 @patch(
     'sdgym.run_benchmark.upload_benchmark_results.DATASET_DETAILS_COLUMNS',
-    ['Dataset', 'Rows', 'Availability', 'Best Model', 'Data Modality'],
+    ['Dataset', 'Total_Num_Rows', 'Availability', 'Best Model', 'Data Modality'],
 )
-def test_get_dataset_details_returns_empty_when_no_datasets_found(mock_dataset_explorer):
+def test_get_dataset_details_returns_empty_when_no_datasets_found(
+    mock_get_available_datasets, mock_dataset_explorer
+):
     """Test the `get_dataset_details` method returns empty DataFrame when no datasets are found."""
     # Setup
     results = pd.DataFrame({
@@ -254,20 +304,26 @@ def test_get_dataset_details_returns_empty_when_no_datasets_found(mock_dataset_e
     public_explorer = Mock()
     private_explorer = Mock()
     mock_dataset_explorer.side_effect = [public_explorer, private_explorer]
-    public_explorer.summarize_datasets.return_value = pd.DataFrame({
-        'Dataset': ['X'],
-        'Rows': [999],
-    })
-    private_explorer.summarize_datasets.return_value = pd.DataFrame({
-        'Dataset': ['Y'],
-        'Rows': [999],
-    })
+    public_explorer._bucket_name = SDV_DATASETS_PUBLIC_BUCKET
+    public_explorer.s3_client = Mock()
+    private_explorer._bucket_name = SDV_DATASETS_PRIVATE_BUCKET
+    private_explorer.s3_client = Mock()
+    mock_get_available_datasets.side_effect = [
+        pd.DataFrame({'dataset_name': ['X'], 'size_MB': [10], 'num_tables': [1]}),
+        pd.DataFrame({'dataset_name': ['Y'], 'size_MB': [10], 'num_tables': [1]}),
+    ]
 
     # Run
     out = get_dataset_details(results, modality, 'access', 'secret')
 
     # Assert
-    assert list(out.columns) == ['Dataset', 'Rows', 'Availability', 'Best Model', 'Data Modality']
+    assert list(out.columns) == [
+        'Dataset',
+        'Total_Num_Rows',
+        'Availability',
+        'Best Model',
+        'Data Modality',
+    ]
     assert out.empty is True
 
 

--- a/tests/unit/run_benchmark/test_upload_benchmark_result.py
+++ b/tests/unit/run_benchmark/test_upload_benchmark_result.py
@@ -2,6 +2,7 @@ import io
 import json
 import re
 from pathlib import Path
+from unittest import mock
 from unittest.mock import Mock, call, mock_open, patch
 
 import pandas as pd
@@ -170,14 +171,12 @@ def test_upload_to_drive_file_not_found(tmp_path):
 
 
 @patch('sdgym.run_benchmark.upload_benchmark_results.DatasetExplorer')
-@patch('sdgym.run_benchmark.upload_benchmark_results._load_dataset_with_client')
-@patch('sdgym.run_benchmark.upload_benchmark_results._get_available_datasets')
 @patch(
     'sdgym.run_benchmark.upload_benchmark_results.DATASET_DETAILS_COLUMNS',
     ['Dataset', 'Total_Num_Rows', 'Availability', 'Best Model', 'Data Modality'],
 )
-def test_get_dataset_details(mock_get_available_datasets, mock_load_dataset, mock_dataset_explorer):
-    """Test the `get_dataset_details` method"""
+def test_get_dataset_details(mock_dataset_explorer):
+    """Test the `get_dataset_details` method."""
     # Setup
     aws_access_key_id = 'access'
     aws_secret_access_key = 'secret'
@@ -199,76 +198,20 @@ def test_get_dataset_details(mock_get_available_datasets, mock_load_dataset, moc
     private_explorer = Mock()
     mock_dataset_explorer.side_effect = [public_explorer, private_explorer]
 
-    public_explorer._bucket_name = SDV_DATASETS_PUBLIC_BUCKET
-    public_explorer.s3_client = Mock()
-    private_explorer._bucket_name = SDV_DATASETS_PRIVATE_BUCKET
-    private_explorer.s3_client = Mock()
-
-    mock_get_available_datasets.side_effect = [
-        pd.DataFrame({
-            'dataset_name': ['A', 'B', 'X'],
-            'size_MB': [1.0, 2.0, 999.0],
-            'num_tables': [1, 1, 1],
-        }),
-        pd.DataFrame({
-            'dataset_name': ['C'],
-            'size_MB': [3.0],
-            'num_tables': [1],
-        }),
+    public_explorer._load_and_summarize_datasets.return_value = [
+        {'Dataset': 'A', 'Total_Num_Rows': 10},
+        {'Dataset': 'B', 'Total_Num_Rows': 20},
     ]
-    mock_load_dataset.side_effect = [
-        ({'table': pd.DataFrame({'col': [1] * 10})}, {'metadata': 1}),
-        ({'table': pd.DataFrame({'col': [1] * 20})}, {'metadata': 1}),
-        ({'table': pd.DataFrame({'col': [1] * 30})}, {'metadata': 1}),
+    private_explorer._load_and_summarize_datasets.return_value = [
+        {'Dataset': 'C', 'Total_Num_Rows': 30},
     ]
 
-    with (
-        patch(
-            'sdgym.run_benchmark.upload_benchmark_results.DatasetExplorer.get_metadata_summary',
-            side_effect=[
-                {
-                    'Total_Num_Columns': 1,
-                    'Total_Num_Columns_Categorical': 1,
-                    'Total_Num_Columns_Numerical': 0,
-                    'Num_Relationships': 0,
-                    'Max_Schema_Depth': 1,
-                },
-                {
-                    'Total_Num_Columns': 1,
-                    'Total_Num_Columns_Categorical': 1,
-                    'Total_Num_Columns_Numerical': 0,
-                    'Num_Relationships': 0,
-                    'Max_Schema_Depth': 1,
-                },
-                {
-                    'Total_Num_Columns': 1,
-                    'Total_Num_Columns_Categorical': 1,
-                    'Total_Num_Columns_Numerical': 0,
-                    'Num_Relationships': 0,
-                    'Max_Schema_Depth': 1,
-                },
-            ],
-        ),
-        patch(
-            'sdgym.run_benchmark.upload_benchmark_results.DatasetExplorer.get_data_summary',
-            side_effect=[{'Total_Num_Rows': 10}, {'Total_Num_Rows': 20}, {'Total_Num_Rows': 30}],
-        ),
-    ):
-        # Run
-        dataset_details = get_dataset_details(
-            results, modality, aws_access_key_id, aws_secret_access_key
-        )
+    # Run
+    dataset_details = get_dataset_details(
+        results, modality, aws_access_key_id, aws_secret_access_key
+    )
 
     # Assert
-    assert mock_get_available_datasets.call_count == 2
-    assert mock_load_dataset.call_count == 3
-    expected = pd.DataFrame({
-        'Dataset': ['A', 'B', 'C'],
-        'Total_Num_Rows': [10, 20, 30],
-        'Availability': ['Public', 'Public', 'Private'],
-        'Best Model': ['TVAESynthesizer', 'GaussianCopulaSynthesizer', 'CTGANSynthesizer'],
-        'Data Modality': [modality, modality, modality],
-    })
     assert mock_dataset_explorer.call_count == 2
     for _call, bucket in zip(mock_dataset_explorer.call_args_list, dataset_bucket):
         assert _call.kwargs == {
@@ -276,6 +219,19 @@ def test_get_dataset_details(mock_get_available_datasets, mock_load_dataset, moc
             'aws_access_key_id': aws_access_key_id,
             'aws_secret_access_key': aws_secret_access_key,
         }
+    public_explorer._load_and_summarize_datasets.assert_called_once_with(
+        modality=modality, datasets=mock.ANY
+    )
+    private_explorer._load_and_summarize_datasets.assert_called_once_with(
+        modality=modality, datasets=mock.ANY
+    )
+    expected = pd.DataFrame({
+        'Dataset': ['A', 'B', 'C'],
+        'Total_Num_Rows': [10, 20, 30],
+        'Availability': ['Public', 'Public', 'Private'],
+        'Best Model': ['TVAESynthesizer', 'GaussianCopulaSynthesizer', 'CTGANSynthesizer'],
+        'Data Modality': [modality, modality, modality],
+    })
     assert_frame_equal(
         dataset_details.sort_values('Dataset').reset_index(drop=True),
         expected.sort_values('Dataset').reset_index(drop=True),
@@ -284,14 +240,11 @@ def test_get_dataset_details(mock_get_available_datasets, mock_load_dataset, moc
 
 
 @patch('sdgym.run_benchmark.upload_benchmark_results.DatasetExplorer')
-@patch('sdgym.run_benchmark.upload_benchmark_results._get_available_datasets')
 @patch(
     'sdgym.run_benchmark.upload_benchmark_results.DATASET_DETAILS_COLUMNS',
     ['Dataset', 'Total_Num_Rows', 'Availability', 'Best Model', 'Data Modality'],
 )
-def test_get_dataset_details_returns_empty_when_no_datasets_found(
-    mock_get_available_datasets, mock_dataset_explorer
-):
+def test_get_dataset_details_returns_empty_when_no_datasets_found(mock_dataset_explorer):
     """Test the `get_dataset_details` method returns empty DataFrame when no datasets are found."""
     # Setup
     results = pd.DataFrame({
@@ -304,14 +257,8 @@ def test_get_dataset_details_returns_empty_when_no_datasets_found(
     public_explorer = Mock()
     private_explorer = Mock()
     mock_dataset_explorer.side_effect = [public_explorer, private_explorer]
-    public_explorer._bucket_name = SDV_DATASETS_PUBLIC_BUCKET
-    public_explorer.s3_client = Mock()
-    private_explorer._bucket_name = SDV_DATASETS_PRIVATE_BUCKET
-    private_explorer.s3_client = Mock()
-    mock_get_available_datasets.side_effect = [
-        pd.DataFrame({'dataset_name': ['X'], 'size_MB': [10], 'num_tables': [1]}),
-        pd.DataFrame({'dataset_name': ['Y'], 'size_MB': [10], 'num_tables': [1]}),
-    ]
+    public_explorer._load_and_summarize_datasets.return_value = []
+    private_explorer._load_and_summarize_datasets.return_value = []
 
     # Run
     out = get_dataset_details(results, modality, 'access', 'secret')

--- a/tests/unit/run_benchmark/test_upload_benchmark_result.py
+++ b/tests/unit/run_benchmark/test_upload_benchmark_result.py
@@ -2,8 +2,7 @@ import io
 import json
 import re
 from pathlib import Path
-from unittest import mock
-from unittest.mock import Mock, call, mock_open, patch
+from unittest.mock import ANY, Mock, call, mock_open, patch
 
 import pandas as pd
 import pytest
@@ -176,7 +175,7 @@ def test_upload_to_drive_file_not_found(tmp_path):
     ['Dataset', 'Total_Num_Rows', 'Availability', 'Best Model', 'Data Modality'],
 )
 def test_get_dataset_details(mock_dataset_explorer):
-    """Test the `get_dataset_details` method."""
+    """Test the `get_dataset_details` method"""
     # Setup
     aws_access_key_id = 'access'
     aws_secret_access_key = 'secret'
@@ -220,10 +219,10 @@ def test_get_dataset_details(mock_dataset_explorer):
             'aws_secret_access_key': aws_secret_access_key,
         }
     public_explorer._load_and_summarize_datasets.assert_called_once_with(
-        modality=modality, datasets=mock.ANY
+        modality=modality, datasets=ANY
     )
     private_explorer._load_and_summarize_datasets.assert_called_once_with(
-        modality=modality, datasets=mock.ANY
+        modality=modality, datasets=ANY
     )
     expected = pd.DataFrame({
         'Dataset': ['A', 'B', 'C'],

--- a/tests/unit/test_dataset_explorer.py
+++ b/tests/unit/test_dataset_explorer.py
@@ -278,6 +278,32 @@ class TestDatasetExplorer:
         assert 'Dataset' in result[0]
         assert set(result[0]) == set(SUMMARY_OUTPUT_COLUMNS)
 
+    @patch('sdgym.dataset_explorer._get_available_datasets')
+    @patch('sdgym.dataset_explorer._load_dataset_with_client')
+    def test__load_and_summarize_datasets_with_datasets(self, mock_load_dataset, mock_get_datasets):
+        """Test that ``_load_and_summarize_datasets`` restricts to the provided dataset list."""
+        # Setup
+        explorer = DatasetExplorer()
+        mock_get_datasets.return_value = pd.DataFrame([
+            {'dataset_name': 'ds1', 'size_MB': 10, 'num_tables': 1},
+            {'dataset_name': 'ds2', 'size_MB': 20, 'num_tables': 2},
+            {'dataset_name': 'ds3', 'size_MB': 30, 'num_tables': 3},
+        ])
+        mock_load_dataset.return_value = (
+            {'table': pd.DataFrame({'x': [1, 2]})},
+            {'tables': {}, 'relationships': []},
+        )
+
+        # Run
+        results = explorer._load_and_summarize_datasets('single_table', datasets=['ds1', 'ds3'])
+
+        # Assert
+        assert mock_load_dataset.call_count == 2
+        loaded_names = [call.kwargs['dataset'] for call in mock_load_dataset.call_args_list]
+        assert set(loaded_names) == {'ds1', 'ds3'}
+        assert [result['Dataset'] for result in results] == ['ds1', 'ds3']
+        assert len(results) == 2
+
     @patch('sdgym.dataset_explorer._validate_modality')
     @patch('sdgym.dataset_explorer.DatasetExplorer._load_and_summarize_datasets')
     def test_summarize_datasets_with_output(

